meteor/client/ui/Prompter/PrompterView.tsx

```diff
@@ -652,47 +657,147 @@ export const Prompter = translateWithTracker<PropsWithChildren<IPrompterProps>,
-				if (top <= windowInnerHeight && bottom >= readPosition) foundPositions.push([top, anchor.id])
-			})
+			// Gather anchors from any text blocks in view:
 
-			// if we didn't find any text, let's use scroll-anchors instead (Segment and Part names)
-			if (foundPositions.length === 0) {
-				Array.from(document.querySelectorAll('.prompter .scroll-anchor')).forEach((anchor) => {
-					const { top } = anchor.getBoundingClientRect()
-					// find a prompter scroll anchor that is just above the prompter read position
-					if (top <= readPosition + margin) foundPositions.push([top, anchor.id])
-				})
+			for (const textAnchor of document.querySelectorAll('.prompter .prompter-line:not(.empty)')) {
+				const { top, bottom } = textAnchor.getBoundingClientRect()
+
+				// Is the text block in view?
+				if (top <= readPosition && bottom > readPosition) {
+					useableTextAnchors.push({ anchorId: textAnchor.id, offset: top })
+				}
 			}
 
-			foundPositions = _.sortBy(foundPositions, (topOffset) => Number(topOffset[0]))
+			// Also use scroll-anchors (Segment and Part names)
+
+			for (const scrollAnchor of document.querySelectorAll('.prompter .scroll-anchor')) {
+				const { top, bottom } = scrollAnchor.getBoundingClientRect()
 
-			if (foundPositions.length > 0) {
-				return foundPositions[foundPositions.length - 1]
+				const distanceToReadPosition = Math.abs(top - readPosition)
+
+				if (top <= windowInnerHeight && bottom > 0) {
+					// If the anchor is in view, use the offset to keep it's position unchanged, relative to the viewport
+					foundScrollAnchors.push({ anchorId: scrollAnchor.id, distanceToReadPosition, offset: top })
+				} else {
+					// If the anchor is not in view, set the offset to null, this will cause the view to
+					// jump so that the anchor will be in view.
+					foundScrollAnchors.push({ anchorId: scrollAnchor.id, distanceToReadPosition, offset: null })
+				}
 			}
-			return null
+
+			// Sort text anchors, topmost first:
+			const sortedTextAnchors = useableTextAnchors.sort((a, b) => {
+				return a.offset - b.offset
+			})
+
+			// Sort, smallest distanceToReadPosition first:
+			const sortedScrollAnchors = foundScrollAnchors.sort((a, b) => {
+				return a.distanceToReadPosition - b.distanceToReadPosition
+			})
+
+			// Prioritize the text anchors, then the scroll anchors:
+			return [...sortedTextAnchors, ...sortedScrollAnchors]
 		}
 
-		private restoreScrollAnchor = (scrollAnchor: ScrollAnchor) => {
-			if (scrollAnchor === null) return
-			const anchor = document.getElementById(scrollAnchor[1])
-			if (anchor) {
-				const { top } = anchor.getBoundingClientRect()
+		private restoreScrollAnchor = (scrollAnchors: ScrollAnchor[] | null) => {
+			if (scrollAnchors === null) return
+			if (!scrollAnchors.length) {
+				if (this.props.prompterData?.segments.length) {
+					// That's wierd, log the issue:
+					logger.error(`No read anchors to use after update`)
```

Comment:
I'm not sure if this is that weird - all it takes is a rundown activation and not getting all segments+parts yet from the publication. `scrollAnchors` will be `null` only on mount, at every update later, it will be an array, and if the data isn't loaded yet, it will be empty, so I think this is to be expected. Anyway, this doesn't provide any context (playlistId, etc.), so maybe it should?